File: cpp/leetcode/leetcode/14_longest_common_prefix.cpp
//
//  14_longest_common_prefix.cpp
//  leetcode
//
//  Created by hatlonely on 16/1/11.
//  Copyright © 2016年 hatlonely. All rights reserved.
//
//  Write a function to find the longest common prefix string amongst an array of strings.
//

#include <iostream>
#include <cassert>
#include <string>
#include <vector>

namespace longest_common_prefix {

class Solution {
public:
    std::string longestCommonPrefix(std::vector<std::string>& strs) {
        if (strs.empty()) {
            return "";
        }
        
        int size = (int)strs.size();
        int length = (int)strs[0].length();
        for (int i = 0; i < length; i++) {
            char ch = strs[0][i];
            for (int j = 1; j < size; j++) {
                if (strs[j][i] != ch) {
                    return strs[0].substr(0, i);
                }
            }
        }
        
        return strs[0];
    }
};

int main(int argc, const char *argv[]) {
    auto test = [=](std::initializer_list<std::string> ls, std::string expected) -> bool {
        Solution solution;
        std::vector<std::string> strs(ls);
        return solution.longestCommonPrefix(strs) == expected;
    };
    
    assert(test({"abc", "abd", "abde", "abcd"}, "ab"));
    assert(test({"ab", "abc", "abd"}, "ab"));
    
    return 0;
}

}

File: go/src/main/14_longest_common_prefix.go
package main

import "fmt"

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	str1 := strs[0]
	for i := 1; i < len(strs); i++ {
		str2 := strs[i]
		idx := 0
		for idx < len(str1) && idx < len(str2) && str1[idx] == str2[idx] {
			idx++
		}
		str1 = str1[:idx]
	}

	return str1
}

func main() {
	for _, unit := range []struct {
		strs   []string
		prefix string
	}{
		{[]string{"abc", "abcd", "ab"}, "ab"},
	} {
		if prefix := longestCommonPrefix(unit.strs); prefix != unit.prefix {
			fmt.Printf("longestCommonPrefix(%v) = %v != %v\n", unit.strs, prefix, unit.prefix)
		}
	}
}
